t/t1092-sparse-checkout-compatibility.sh

```diff
@@ -975,6 +975,21 @@ test_expect_success 'sparse index is not expanded: sparse-checkout' '
 	ensure_not_expanded sparse-checkout set
 '
 
+test_expect_success 'sparse index is not expanded: update-index' '
+	init_repos &&
+
+	echo "test" >sparse-index/README.md &&
+	echo "test2" >sparse-index/a &&
+	rm -f sparse-index/deep/a &&
+
+	ensure_not_expanded update-index --add README.md &&
+	ensure_not_expanded update-index a &&
+	ensure_not_expanded update-index --remove deep/a &&
+
+	rm -f sparse-index/README.md sparse-index/a &&
+	ensure_not_expanded update-index --add --remove --again
+'
```

Comment:
There are a lot of code changes in this commit. I wonder if it is possible to add each `ensure_not_expanded` line from this test in a different commit, and only include the code changes necessary to make that line pass. It would assist with upstream review, for sure.

To make that work, you might need to reorder your previous test-expansion commit to be _after_ all of the sparse index integrations. Otherwise, maybe _they_ will start failing after `command_requires_full_index` is disabled but the integration isn't complete. Alternatively, could those tests be sprinkled with the code changes as they start working with the sparse index?

Reply:
I attempted to break things up a bit more, either by adding "temporary" `ensure_full_index(...)` calls or overall reducing the changes by removing some (seemingly?) unnecessary `ensure_full_index(...)` or `S_ISSPARSEDIR(...)` checks. Let me know if you think they should be broken up more, though.